8285295: Need better testing for IdentityHashMap

stuart-marks opened the pull request:
Basic but fairly comprehensive set of tests for `IdentityHashMap`. The patch in the bug report that breaks `IdentityHashMap` now causes several cases in this new test to fail. There's more that could be done, but the new tests cover most of the core functions of `IdentityHashMap`. Unfortunately it seems difficult to merge this with the existing, comprehensive Collections tests (e.g., MOAT.java) because those tests implicity rely on `equals()`-based contract instead of the special-purpose `==`-based contract used by `IdentityHashMap`.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issue
 * [JDK-8285295](https://bugs.openjdk.java.net/browse/JDK-8285295): Need better testing for IdentityHashMap


### Reviewers
 * [Jaikiran Pai](https://openjdk.java.net/census#jpai) (@jaikiran - Committer) ⚠️ Review applies to [4bb25edf](https://git.openjdk.java.net/jdk/pull/8354/files/4bb25edf8d47e117a8ed32f9facb51b2dbb0b04d)
 * [Lance Andersen](https://openjdk.java.net/census#lancea) (@LanceAndersen - **Reviewer**) ⚠️ Review applies to [4bb25edf](https://git.openjdk.java.net/jdk/pull/8354/files/4bb25edf8d47e117a8ed32f9facb51b2dbb0b04d)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8354/head:pull/8354` \
`$ git checkout pull/8354`

Update a local copy of the PR: \
`$ git checkout pull/8354` \
`$ git pull https://git.openjdk.java.net/jdk pull/8354/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8354`

View PR using the GUI difftool: \
`$ git pr show -t 8354`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8354.diff">https://git.openjdk.java.net/jdk/pull/8354.diff</a>

</details>

liach commented:
Just curious, will entrySet/keySet/values tests be here or in a separate file? They are identity-based collections as well, as shown by their contains/removeAll/retainAll.

stuart-marks commented:
Probably the same file. It would be nice to be able to re-use some general Set and Collection tests, but the identity semantics make this difficult to do without extensive refactoring.

liach commented on test/jdk/java/util/IdentityHashMap/Basic.java:
```diff
@@ -38,33 +43,84 @@
+    @SafeVarargs
+    private <E> void checkContents(Collection<E> c, BiPredicate<E,E> p, E... given) {
+        @SuppressWarnings("unchecked")
+        E[] contents = (E[]) c.toArray();
```
Maybe worthy to note that entry set toArray may return entries different from what the iterator returns, and some predicates may thus fail.

stuart-marks replied:
Yeah I could add a note to caution against checking identity of `Entry` instances. That doesn't occur anywhere in these tests, does it?

liach commented on test/jdk/java/util/IdentityHashMap/Basic.java:
```diff
@@ -38,33 +43,84 @@
+        @SuppressWarnings("unchecked")
+        E[] contents = (E[]) c.toArray();
+
+        assertEquals(c.size(), given.length);
```
I believe testng has the expected values in front in the `assertEquals` methods, as embodied in the exception messages, so this should be `assertEquals(given.length, c.size());`. Applies to other places.

stuart-marks replied:
No, TestNG is `assertEquals(actual, expected)` which is irritatingly the opposite of JUnit.

https://github.com/cbeust/testng/blob/master/testng-asserts/src/main/java/org/testng/asserts/Assertion.java#L151

This will make things quite tedious when/if we convert to JUnit.

ExE-Boss replied:
There’s a reason why I prefer using [AssertJ], where the calls are:
```java
Assertions.assertThat(actual)
	.isEqualTo(expected);
```

[AssertJ]: https://assertj.github.io/doc

jaikiran commented on test/jdk/java/util/IdentityHashMap/Basic.java:
```diff
@@ -0,0 +1,546 @@
+    @Test
+    public void testKeySetNoRemove() {
+        Set<Box> keySet = map.keySet();
+        keySet.remove(new Box(k1a));
```
Should we assert here that this returns `false`?

jaikiran commented on test/jdk/java/util/IdentityHashMap/Basic.java:
```diff
@@ -0,0 +1,546 @@
+    @Test
+    public void testKeySetRemove() {
+        Set<Box> keySet = map.keySet();
+        keySet.remove(k1a);
```
Similarly should we assert `true` here?

jaikiran commented on test/jdk/java/util/IdentityHashMap/Basic.java:
```diff
@@ -0,0 +1,546 @@
+        checkElements(keySet, k1b, k2);
+        checkEntries(map.entrySet(), entry(k1b, v1b),
+                                     entry(k2, v2));
+    }
```
Would an additional check `assertFalse(map.equals(map2));` be useful here (and other similar tests where we do "remove").

stuart-marks replied:
I don't know if you noticed that previous versions checked the map's contents with `map.equals(map2)` and either asserted true or false depending on the situation. I pulled most of those out when I added `checkEntries`. The reason is that `checkEntries` and `checkElements` are supposed to check the exact contents of the map or the collection, and they fail if there is a missing, different, or extra entry or element. If that's true we don't need to test `map.equals`. I don't think it calling `map.equals` adds any value or does any checking that the `checkX` methods don't already do.

Of course this relies on `checkEntries` and `checkElements` to do their jobs properly, so we should make sure they do!

And also we need to test that the `equals` method is doing its job as well. There are a couple tests for it already, and they test at least the basic cases. But it's possible I might have missed something.

In any case, if we believe the `checkX` methods are sufficient, and if we believe that the tests for `equals` are also sufficient, then I don't think we need to add assertions about `equals` in any tests other than for `equals` itself.

jaikiran commented on test/jdk/java/util/IdentityHashMap/Basic.java:
```diff
@@ -0,0 +1,546 @@
+    public void testPutNew() {
+        Box newKey = new Box(k1a);
+        Box newVal = new Box(v1a);
+        map.put(newKey, newVal);
```
Should we capture the return value and assert that it is `null`?

jaikiran commented on test/jdk/java/util/IdentityHashMap/Basic.java:
```diff
@@ -0,0 +1,546 @@
+    @Test
+    public void testPutOverwrite() {
+        Box newVal = new Box(v1a);
+        map.put(k1a, newVal);
```
Should we capture the return value and assert that it's not null and is identity equal to `v1a`?

stuart-marks replied:
Yes, good point, I'll add checks of the return values in the appropriate places. There are several, including `remove`, `computeX`, `put`, etc.

stuart-marks commented:
I've added checking of return values for I think everything that has a significant return value. I've elected to store the return value in a local variable and add a separate assert line. For example, instead of

    assertNull(map.put(newKey, newVal));

I've done

    Box r = map.put(newKey, newVal);
    assertNull(r);

The reason is that I think it separates the test setup/action from the test assertions. I tried it the first way, but it felt like the lack of this separation made things messy.

I've also added a couple more tests over `equals` and added more asserts over `equals` to the `keySet` and `entrySet` view sets. (Note, the `values` collection is just a `Collection` and thus doesn't have a defined notion of `equals`.) The testing of the view collections could probably be made more comprehensive, but I think what's here is a good start.

Please take a look.

jaikiran commented on test/jdk/java/util/IdentityHashMap/Basic.java:
```diff
@@ -0,0 +1,628 @@
+        boolean[] called = new boolean[1];
+        Box newKey = new Box(k1a);
+        Box newVal = new Box(v1a);
+        Box r = map.computeIfAbsent(newKey, k -> { called[0] = true; return newVal; });
```
More of a curiosity than a review comment - I see that various places in this PR use a boolean array with one element instead of just a boolean. Is that a personal coding preference or is there something more to it?

stuart-marks replied:
Yes, that's it; you can't mutate a local variable captured by a lambda.

jaikiran replied:
Thank you @liach and Stuart. I had overlooked the lambda aspect of this.

stuart-marks commented on test/jdk/java/util/IdentityHashMap/Basic.java:
```diff
@@ -0,0 +1,628 @@
+// as its logic for testing collections equality is suspect. Use checkContents() to assert
+// that a map's entrySet contains exactly the expected mappings. In most cases, keys and
+// values should be compared by identity, not equality. However, the identities of Map.Entry
+// instances from an IdentityHashSet are not guaranteed; the keys and values they contain
```
Yeah that comment was poorly worded, and indeed I meant the IdentityHashMap's entrySet and not IdentityHashSet. I've reworded it. I was trying to make a statement about exactly what needs to be compared if you want to make assertions about two maps being "equal" in the sense of IdentityHashMap behaving correctly. Specifically, given two maps `m1` and `m2`, clearly we don't want either of

    assertSame(m1, m2);
    assertSame(m1.entrySet(), m2.entrySet());

Instead, we want the `Map.Entry` instances to "match". However, given two entries `entry1` and `entry2` that are supposed to match, we also do not want

    assertSame(entry1, entry2);

Instead, we want

    assertSame(entry1.getKey(), entry2.getKey());
    assertSame(entry1.getValue(), entry2,getValue());

The `checkEntries` method goes to some length to get this right.

stuart-marks commented:
I've added a few more assertions to cover `equals` of `entrySet` and `keySet` which I think were missing some cases. (Note that `values` returns a `Collection` which has no notion of equality.) I think this is ready now.

jaikiran reviewed (approved):
Hello Stuart, these changes look good to me.

ExE-Boss commented:
As I said in [GH‑6532]:
> There should probably be something like [test/jdk/java/util/Collections/Wrappers.java] to check that `IdentityHashMap` overrides all `default` methods from `java.util.Map` (with `remove(K, V)` and `replace(K, V, V)` depending on [GH‑8259]).

[GH‑6532]: https://github.com/openjdk/jdk/pull/6532#issuecomment-1104709021
[GH‑8259]: https://github.com/openjdk/jdk/pull/8259
[test/jdk/java/util/Collections/Wrappers.java]: https://github.com/openjdk/jdk/blob/master/test/jdk/java/util/Collections/Wrappers.java

stuart-marks commented:
Thanks Jaikiran. Could I have a Reviewer take a look at this please?

stuart-marks commented:
> There should probably be something like [test/jdk/java/util/Collections/Wrappers.java](https://github.com/openjdk/jdk/blob/master/test/jdk/java/util/Collections/Wrappers.java)

Maybe. The intent of the test is fine, which is to ensure that a default method doesn't get added that breaks the invariants of the wrapper. One problem is that it tests only the default methods of `Collection` and not of the other collections interfaces. Another is that "override all default methods" isn't exactly the right criterion; instead, the criterion should be "override all default methods that would otherwise break this collection's invariants." It would be nice if such a test could be written, but as it stands I think that `Wrappers.java` test is too simplistic.

XenoAmess commented:
> It would be nice if such a test could be written, but as it stands I think that `Wrappers.java` test is too simplistic.

Would adding `Wrappers.java` a method-name white-list mechanism suitable in this situation?

ExE-Boss commented:
> > It would be nice if such a test could be written, but as it stands I think that `Wrappers.java` test is too simplistic.
> 
> Would adding `Wrappers.java` a method-name white-list mechanism suitable in this situation?

It should really be a method name and type white‑list.

LanceAndersen reviewed (approved):
I think you have done a nice job on the coverage.

It would be nice for future maintainers if you consider adding comments for all of the tests vs. a subset prior to pushing

stuart-marks commented:
Thanks @LanceAndersen I've added some comments.

stuart-marks commented:
/integrate
